egui/src/widgets/text_edit.rs

```diff
@@ -432,12 +502,34 @@ impl<'t> TextEdit<'t> {
                             && text_to_insert != "\r"
                         {
                             let mut ccursor = delete_selected(text, &cursorp);
+
                             insert_text(&mut ccursor, text, text_to_insert);
                             Some(CCursorPair::one(ccursor))
                         } else {
                             None
                         }
                     }
+                    Event::Key {
+                        key: Key::Tab,
+                        pressed: true,
+                        ..
+                    } => {
+                        if multiline {
+                            let mut ccursor = delete_selected(text, &cursorp);
+
+                            if ui.memory().has_lock_focus(id) {
+                                if tab_as_spaces {
+                                    insert_text(&mut ccursor, text, "    ");
+                                } else {
+                                    insert_text(&mut ccursor, text, "\t");
+                                }
+                            }
```

Comment:
It would be great to also implement shift-tab (delete leading tab/spaces from the current line). A bit more work though due to the complicated text cursor system.

Reply:
With this commit https://github.com/emilk/egui/pull/246/commits/abb8b538c4ddbb2c554bc79be491fd9f5ddd66d8 the bulk of the work should be done. I'm going to iron out some things later. Stuff like what to do when we have a selection and we press `shift`+`tab`.

Reply:
After extensive testing I concluded that what I proposed above is garbage. Now I try to wrap my head around the cursor system while starting kind of from scratch. Nevertheless I won't stop until I get it done.

Reply:
Last push contains indentation management using `tab`, `shift`+`tab`, `ctrl`+`[` and `shift`+`]`.
What may need to still be polished is selection update when adding or removing indentation or when converting from spaces to tabs.